src/pkg/auth/client.go

```diff
@@ -0,0 +1,249 @@
+	return func(vo *VerifyOptions) {
+		vo.refresh = refresh
+	}
+}
+
+type Tokens struct {
+	AccessToken  string `json:"access_token"`
+	RefreshToken string `json:"refresh_token"`
+	// ExpiresIn   int    `json:"expires_in"`
+	// Scope       string `json:"scope"`
+}
+
+type VerifyResult struct {
+	*Tokens
+}
+
+type Client interface {
+	/**
+	 * Start the autorization flow.
+	 * This returns a redirect URL and a challenge that you need to use later to verify the code.
+	 */
+	Authorize(redirectURI string, response ResponseType, opts ...AuthorizeOption) (AuthorizeResult, error)
+	/**
+	 * Exchange the code for access and refresh tokens.
+	 */
+	Exchange(code string, redirectURI string, verifier string) (ExchangeSuccess, error)
+	/**
+	 * Refreshes the tokens if they have expired. This is used in an SPA app to maintain the
+	 * session, without logging the user out.
+	 */
+	Refresh(refresh string, opts ...RefreshOption) (RefreshSuccess, error)
+	/**
+	 * Verify the token in the incoming request.
+	 */
+	Verify(token string, opts ...VerifyOption) (VerifyResult, error)
+}
+
+type client struct {
+	clientID string
+	issuer   string
+}
+
+func NewClient(clientID, issuer string) *client {
+	return &client{
+		clientID: clientID,
+		issuer:   issuer,
+	}
+}
+
+func (c client) Authorize(redirectURI string, response ResponseType, opts ...AuthorizeOption) (AuthorizeResult, error) {
+	var as AuthorizeOptions
+	for _, o := range opts {
+		o(&as)
+	}
+
+	result, _ := url.Parse(c.issuer + "/authorize")
+	state := uuid.NewString()
+	values := url.Values{
+		"client_id":     {c.clientID},
+		"state":         {state},
+		"redirect_uri":  {redirectURI},
+		"response_type": {string(response)},
+		// "scope":         {"read:org user:email"}, TODO: add scope AuthorizeOption
+		// "login":         {";TODO: from state file"},
+	}
+	if as.provider != "" {
+		values.Set("provider", as.provider)
+	}
+	var verifier string
+	if as.pkce && response == "code" {
+		pkce, err := GeneratePKCE(64)
+		if err != nil {
+			return AuthorizeResult{}, err
+		}
+		values.Set("code_challenge_method", string(pkce.Method))
+		values.Set("code_challenge", pkce.Challenge)
+		verifier = pkce.Verifier
+	}
+	result.RawQuery = values.Encode()
+	return AuthorizeResult{
+		state:    state,
+		verifier: verifier,
+		url:      *result,
+	}, nil
+}
+
+/**
+ * Exchange the code for access and refresh tokens.
+ */
+func (c client) Exchange(code string, redirectURI string, verifier string) (ExchangeSuccess, error) {
+	body := url.Values{
+		"code":          {code},
+		"redirect_uri":  {redirectURI},
+		"grant_type":    {"authorization_code"},
+		"client_id":     {c.clientID},
+		"code_verifier": {verifier},
+	}
+
+	resp, err := http.PostForm(c.issuer+"/token", body)
+	if err != nil {
+		return ExchangeSuccess{}, err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return ExchangeSuccess{}, errors.New("invalid authorization code: " + resp.Status)
+	}
```

Comment:
This feels like a greedy error message.

The server probably returns a very specific status code if the authorization code is truly invalid

Could we be more specific by handling non-OK status codes with something like this?

```suggestion
	if resp.StatusCode == http.StatusBadRequest {
		// not sure if HTTP 400 is the expected code, but seems reasonable
		return RefreshSuccess{}, errors.New("invalid authorization code")
	} else if resp.StatusCode != http.StatusOK {
		// a more generic error message here
		return RefreshSuccess{}, errors.new("failed to exchange authorization code for tokens" + resp.Status)
	}
```

Reply:
It's not using HTTP status codes, but `error` in the JSON. Will updated.

Reply:
This has not been resolved. I will still see an error message that says "invalid authorization code:" if the server returns HTTP 5xx.

Reply:
Done